src/perform_step/adams_bashforth_moulton_perform_step.jl

```diff
@@ -772,3 +772,138 @@ end
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack dts,g,ϕ_n,ϕstar_n,ϕstar_nm1,order,rk4constcache = cache
+  k1 = integrator.fsalfirst
+  if integrator.u_modified
+    cache.step = 1
+  end
+  k = cache.step
+  if k == 1
+    dts[1] = dt
+    cache.step += 1
+  elseif k == 2
+    dts[2] = dts[1]
+    dts[1] = dt
+    cache.step += 1
+  elseif k == 3
+    dts[3] = dts[2]
+    dts[2] = dts[1]
+    dts[1] = dt
+    cache.step += 1
+  elseif k == 4
+    dts[4] = dts[3]
+    dts[3] = dts[2]
+    dts[2] = dts[1]
+    dts[1] = dt
+    cache.step += 1
+  else
+    dts[5] = dts[4]
+    dts[4] = dts[3]
+    dts[3] = dts[2]
+    dts[2] = dts[1]
+    dts[1] = dt
+  end
+  ϕ_and_ϕstar!(cache,k1,k)
+  cache.ϕstar_nm1 .= ϕstar_n
+  if k == 1 || k == 2 || k == 3 || k == 4
+    perform_step!(integrator, rk4constcache)
+  else
+    g_coefs!(cache,k)
+    u = uprev
+    for i = 1:k
+        u += dt * g[i] * ϕstar_n[i]
+    end
+    if integrator.opts.adaptive
+      utilde = dt * g[k] * ϕstar_n[k]
+      atmp = calculate_residuals(utilde, uprev, u, integrator.opts.abstol, integrator.opts.reltol,integrator.opts.internalnorm)
+      integrator.EEst = integrator.opts.internalnorm(atmp)
+    end
+    integrator.fsallast = f(u, p, t+dt)
+    integrator.k[1] = integrator.fsalfirst
+    integrator.k[2] = integrator.fsallast
+    integrator.u = u
+  end
+end
+
+function initialize!(integrator,cache::VCAB5Cache)
+  @unpack fsalfirst,k4 = cache
+  integrator.fsalfirst = fsalfirst
+  integrator.fsallast = k4
+  integrator.kshortsize = 2
+  resize!(integrator.k, integrator.kshortsize)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+  integrator.f(integrator.fsalfirst,integrator.uprev,integrator.p,integrator.t) # pre-start FSAL
+end
+
+@muladd function perform_step!(integrator,cache::VCAB5Cache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack k4,dts,g,ϕ_n,ϕstar_n,ϕstar_nm1,order,atmp,utilde,rk4cache = cache
+  k1 = integrator.fsalfirst
+  if integrator.u_modified
+    cache.step = 1
+  end
+  k = cache.step
+  if k == 1
+    dts[1] = dt
+    cache.step += 1
+  elseif k == 2
+    dts[2] = dts[1]
+    dts[1] = dt
+    cache.step += 1
+  elseif k == 3
+    dts[3] = dts[2]
+    dts[2] = dts[1]
+    dts[1] = dt
+    cache.step += 1
+  elseif k == 4
+    dts[4] = dts[3]
+    dts[3] = dts[2]
+    dts[2] = dts[1]
+    dts[1] = dt
+    cache.step += 1
+  else
+    dts[5] = dts[4]
+    dts[4] = dts[3]
+    dts[3] = dts[2]
+    dts[2] = dts[1]
+    dts[1] = dt
+  end
+  ϕ_and_ϕstar!(cache,k1,k)
+  for i in eachindex(ϕstar_n)
+    cache.ϕstar_nm1[i] .= ϕstar_n[i]
+  end
+  if k == 1 || k == 2 || k == 3 || k == 4
+    rk4cache.fsalfirst .= k1
+    perform_step!(integrator, rk4cache)
+    integrator.fsallast .= rk4cache.k
+  else
+    g_coefs!(cache,k)
+    @. u = uprev
+    for i = 1:k
+      @. u += dt * g[i] * ϕstar_n[i]
+    end
+    f(k4,u,p,t+dt)
```

Comment:
Extra function evaluation?

Reply:
Nevermind. I see that's the FSAL variable, so it's just setting the future past step.